File: Data Structures & Algorithms/search-2d-matrix/submission-2.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def search(nums: List[int], target: int) -> int:
            l = 0
            r = len(nums)-1
            while l<=r:
                m = (r+l)//2
                if nums[m]<target:
                    l = m+1
                elif nums[m]>target:
                    r = m-1
                else:
                    return m
            return -1
        l = 0
        r = len(matrix)-1
        while l <= r:
            m = (r+l)//2
            nums = matrix[m]
            if matrix[m][-1]<target:
                l = m+1
            elif matrix[m][0]>target:
                r = m-1
            else:
                break
        return search(nums, target) != -1

File: Data Structures & Algorithms/search-insert-position/submission-2.go
func searchInsert(nums []int, target int) int {

	i, j := 0, len(nums)-1
	m := 0
	for i <= j {
		m = i + (j-i)/2
		if nums[m] < target {
			i = m + 1
		} else if nums[m] > target {
			j = m - 1
		} else {
			i = m
			break
		}
	}

	return i

}
